### DISCUSSION TITLE
Pattern to call a dag with different parameters?

### DISCUSSION THREAD

**Discussion by match-gabeflores** [apache/airflow#20937] [Q&A]
I have a dag that calls a stored procedure and it runs daily at 7pm.
`exec dbo.MyProc @proc_param = 'Client1`

The requirement is to add another dag that is identical to it except it's for a different client and thus the proc needs to be something like this:
`exec dbo.MyProc @proc_param = 'Client2`

The proc for Client2 needs to run on a completely different schedule (weekly)

This is a simplified case.. my real case is more complicated and it's code/logic that i don't want to duplicate across several dags. Is the best approach to have three files: 

- a controller dag with daily schedule that triggers (using `TriggerDagRunOperator`)the dag for client1 by passing in conf={"proc_param": "Client1"}
- a controller dag with weekly schedule that triggers the dag for client2 by passing in conf={"proc_param": "Client2"}
- the main dag with the code to run the proc. It'll use something like dag_run.conf.get('proc_param') to get the config value that was passed in

Sometimes the schedule can be the same, in this case I think I would be fine with just one controller dag. But unfortunately, it seems like if I had 10 clients with different schedules, I would need 10 different controller dags.

**Comment by jedcunningham** [comment]
One thing to keep in mind is that DAGs are just Python. You can write a "task generator" to build the task for different clients. I'd explore that possibility before I went down the TriggerDagRunOperator and params path, personally.

Here is an example DAG that shows what I mean (it can be in another file too): https://github.com/apache/airflow/blob/cdb120de2403f5a21aa6d84d10f68c1b7f086aba/airflow/example_dags/example_skip_dag.py#L40-L60

One should be aware that doing anything expensive in top level DAG code can be problematic, so that does limit your options a bit. More details can be found at:
https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code
